### ISSUE TITLE
Dynamic height / scrollbars

### ISSUE THREAD

**Issue by patrik-u** [brimdata/react-arborist#86]
The scrollbar is a bit ugly. Is it possible to have the Tree adjust its height dynamically instead of having it fixed and let the parent container handle the scrollbars? Or alternatively be able to style the scrollbars so that it e.g. they are only visible on mouse hover?

**Comment by patrik-u** [comment]
I managed to solve this by calculating height based on number of nodes in the tree and adjusting the height of the Tree component accordingly.

**Comment by jameskerr** [comment]
I usually make the height of the tree the same height as it's parent using a ResizeObserver on the parent, the passing the width/height to the tree.

If you make the tree height the same height as all the nodes, you will loose the row virtualization that makes it render fast. For example, if you have 10,000 nodes and the tree height is set to 10,000 x rowHeight,  you'll be rendering all 10,000 of them in the tree. Whereas if the height is only 500, you'll only be rendering 20-30.

I would welcome an example where the scrollbar is custom styled though.

**Comment by Marius-AdamDT** [comment]
@jameskerr do you have a current example with the above? I'm also trying to set the height of the tree dynamically as they expand

**Comment by martinburford** [comment]
Hi @jameskerr Could you please post some code example of how you achieve this? I've been wrestling with this for around 8 hours now, and see some strange results when I try to do this. I've written a few React hooks, tried native JavaScript, but I seem to be get inconsistent results (using ResizeObserver). I actually also get the same inconsistencies when writing regular event listeners on the window.resize as well. Basically, the height of the tree seems to be applied slightly higher than the value ResizeObserver surfaces. And in an inconsistent way. I tried debouncing, but there was still the same issue.

Hence, if you could provide some help with this (via the provision of code for how you use ResizeObserver), that would be appreciated.

Many thanks

**Comment by jameskerr** [comment]
Debugging stuff like this is hard. Dang. I’ll give this a more thorough read and post and example, but I’ve had to make the parent component have styles like min-height: 0 and flex: 1

**Comment by AtiqGauri** [comment]
hey @jameskerr  can you provide an example, better in documentation.

**Comment by emmaivarsson** [comment]
I would also love some better documentation on this.

**Comment by emmaivarsson** [comment]
It would be really nice to have the possibility to send a maxHeight and then for the table to be of height auto otherwise

**Comment by elmcapp** [comment]
Is it posible to use the height of the parent. I am not setting any height but my issue is that there is a lot of space at the bottom of the tree. 

Example my parent height is 252 however the tree is 500. Yes I could set the heigh manually but my parent window is drainable and resizable. 

parent: PanelWindow.tsx
````
return (
    <Draggable
      handle=".handle"
      grid={[25, 25]}
      scale={1}
      defaultPosition={{ x: positionX, y: positionY }}
    >
      <div className={className} style={{ width, height }}>
        {isVisible && (
          <ResizableBox
            style={{ height: initialHeight, width: initialWidth }}
            width={width}
            height={height}
            minConstraints={[250, 250]}
            maxConstraints={[window.innerWidth * 0.9, window.innerHeight * 0.9]}
            resizeHandles={["se", "s", "e"]}
            onResize={handleResize}
          >
            <Card style={{ width, height, boxSizing: "content-box" }}>
              <CardDescription className="handle text-center" ref={headerRef}>
                <XCircle
                  className="absolute top-0 left-0"
                  size={"15px"}
                  onClick={handleDismiss}
                />
                {panelTitle}
              </CardDescription>
              <div
                style={{
                  height: `calc(100% - ${headerHeight}px)`,
                  overflow: "auto",
                }}
              >
                {children}
              </div>
            </Card>
          </ResizableBox>
        )}
      </div>
    </Draggable>
  );
````

ScenePanel.tsx
````
function ScenePanel() {
  const viewRef = useRef<HTMLDivElement>(null);
  const getViewHeight = () => viewRef.current?.clientHeight || 0;

  useEffect(() => {
    // You can now use getViewHeight() to get the height of the view
    const height = getViewHeight();
    console.log("View height:", height);
  }, []);
  
  return (
    <div>
      <PanelWindow
        isVisible={true}
        className="scene"
        panelTitle="Scenes"
        onDismiss={() => {}}
        positionY={0}
        positionX={0}
      >
        <div ref={viewRef}>
          <Tree initialData={demoData}>{Node}</Tree>
        </div>
      </PanelWindow>
    </div>
  );
}
````

The ScenePanel size is showing 500 while the parent is only 252.

**Comment by jameskerr** [comment]
Yes, I usually make the parent expand to fill it's parent (either with flex: 1 or the css below).
```css
height: 100%;
min-height: 0;
```

Then I add a ref to it with this package https://github.com/ZeeCoder/use-resize-observer

That hook will return the height and width of the parent whenever it changes. You then pass these numbers to the Tree.

```
const { ref, width, height } = useResizeObserver<HTMLDivElement>();

<div className="parent" ref={ref}>
  <Tree height={height} width={width} />
</div>
```

**Comment by krillboi** [comment]
Also having some trouble getting this to work the way I want. Is it possible to disable scrolling for the tree and have another component take over that role? For example we have a ScrollArea implementation from Radix UI that we could use. Or would this screw with the virtualization?

I have tried getting the same behavior as the Cities example, where horizontal scrolling is disabled and text has ellipsis overflow instead by using the FillFlexPanel, but I cannot get the horizontal scrollbar to be hidden with `overflow-x: "hidden"` since the root element of the tree seems to have `overflow: "auto"` hard-coded into it?

**Comment by jameskerr** [comment]
Ok so in that case, you'll need to set overflow hidden on your Node renderer. Handle the ellipses in the Node Renderer as well.

The tree component should fill its container, and the tree wrapper should have overflow scroll, but then the nodes should be width: 100% with overflow: hidden.

I've never used custom scrollbars in my app, so if you get it working and want to throw up an example, I could publish that in an FAQ somewhere.

**Comment by krillboi** [comment]
It looks like it could be possible to use custom scrollbars if the `outerElementType` of the `react-window` implementation was able to be swapped out: https://github.com/bvaughn/react-window/issues/110#issuecomment-469061213.

In the example they are using `react-custom-scrollbars` so I don't see why another component such as `ScrollArea` from Radix shouldn't work.

**Comment by jameskerr** [comment]
Cool! Thanks for researching that.

**Comment by MalyugaSensei** [comment]
> Да, я обычно расширяю родительский элемент, чтобы заполнить его (либо с помощью flex: 1, либо с помощью CSS ниже).
> 
> ```css
> height: 100%;
> min-height: 0;
> ```
> 
> Затем я добавляю к нему ссылку с помощью этого пакета [https://github.com/ZeeCoder/use-resize-observer .](https://github.com/ZeeCoder/use-resize-observer)
> 
> Этот крючок будет возвращать высоту и ширину родительского элемента всякий раз, когда он изменяется. Затем вы передаете эти числа в Дерево.
> 
> ```js
> const { ref, width, height } = useResizeObserver();
> 
> <div className="parent" ref={ref}>
>   <Tree height={height} width={width} />
> </div>
> ```
I used your example as a solution to the problem, but the height of the tree is still 500px (I think this is the default)
```
const {ref, height } = useResizeObserver()

return (
        <div ref={ref} style={{minHeight: "0", height: "100%"}}>
                <Tree data={structure}
                    indent={24}
                    height={height}
                    width="auto"
                    rowHeight={66}
                    idAccessor='uuid'
                >
                    {children}
                </Tree>
        </div>
    )
```
![image](https://github.com/brimdata/react-arborist/assets/31480571/ca47b101-eb3a-4dc7-8bae-84bae09b3cba)

**Comment by jameskerr** [comment]
You'll need to make sure that the parent is filling its container. It needs to have flex: 1; on it or grid-template-row 1fr or something to make it fill the container.

**Comment by MalyugaSensei** [comment]
Yes, I tried using flex, but it also doesn't give any result. I tried to reproduce this code in codesandbox to make sure on whose side the problem is. Maybe I don't understand correctly how flex works
https://codesandbox.io/s/nostalgic-carlos-2l38qd?file=/src/App.js

**Comment by JohnAmadeo** [comment]
For those using Tailwind, I was able to apply custom styles to the scrollbar with this nifty Tailwind plugin: https://github.com/adoxography/tailwind-scrollbar

Add the Tailwind classes to the <Tree> component

**Comment by theearlofsandwich** [comment]
I had issues using useResizeObserver where it would resize the height down the way, but if you resized back up it wouldn't fire the resize events. Managed to fix it by setting  height = {height -1 }  in the tree component. the - 1 was just enough to stop the tree from locking the height!

**Comment by danielgormly** [comment]
I did get dynamic height working but it wasn't easy.

The principal is that you want:
1) The list to expand to fit the contents of the tree.
2) but to not exceed the contents of some container it is contained within.
3) The virtual list inside to continue to fill all available space.

Essentially we want our tree height to be equal to `min(container height, total calculated tree height)`.

One limitation that is difficult to get around is that we cannot access the calculated tree height easily without changing react-arborist's code. However we can get to it by polling for it within the Tree component - unfortunately there is no other good way to figure out WHEN this becomes available on the DOM. This would be a lot easier with code that let's you subscribe to changes on the total calculated tree height!

I will extrapolate from my working version to show the principals at work:

First set up the DOM & some refs.
```
const screenRef = useRef(); // where you want to fill the available space of
const containerRef = useRef(); // A div immediately above the tree (may be able to skip this actually (?) and do everything off the screenRef)
const [treeHeight, setTreeHeight] = useState();
<div css={css`height: 100vh; display: flex; flex-direction: column}` ref={screenRef}>
  <div css={css`height: fit-content;`} ref={containerRef}>
    <Tree height={treeHeight} data={...}>
  </div>
</div>
```

Then let's introduce the polling code.
```
// 500 = arbitrary starting height
const [treeContentsHeight, setTreeContentsHeight] = useState(500);
const [screenHeight, setScreenHeight] = useState(500);
useEffect(() => {
    // This observer 
    const observer = new ResizeObserver(entries => {
      const height = entries[0].contentRect.height;
      setTreeContentsHeight(height);
    });
    // This tree has the correct total height calculated so we want to watch it using ResizeObserver
    let innerTreeElement;
    let pollId;
    const poll = () => {
      pollId = setTimeout(() => {
        innerTreeElement = containerRef.current?.firstChild?.firstChild?.firstChild;
        if (innerTreeElement) {
          clearTimeout(pollId);
          observer.observe(innerTreeElement);
        } else {
          pollId = poll();
        }
      }, 30);
    };
    poll();
    return () => {
      observer.disconnect();
      clearTimeout(pollId);
    };
  }, []);
```

Finally, whenever the treeContentsHeight is updated, we can set the containerHeight, and make sure it's updated on resize as well.
```
  useEffect(() => {
    const resizeContainer = (): void => {
      const screenRefHeight = screenRefHeight.current.getBoundingClientRect().height;
      setTreeHeight(Math.min(screenRefHeight, treeContentsHeight));
    };
    resizeContainer();
    window.addEventListener('resize', resizeContainer);
    return () => window.removeEventListener('resize', resizeContainer);
  }, [treeContentsHeight]);
```

**Comment by MalyugaSensei** [comment]
> Мне удалось заставить работать динамическую высоту, но это было непросто.
> 
> Главное, что вы хотите:
> 
> 1. Список, который нужно расширить, чтобы он соответствовал содержимому дерева.
> 2. но не превышать содержимое какого-либо контейнера, внутри которого оно содержится.
> 3. Виртуальный список внутри продолжит заполнять все доступное пространство.
> 
> По сути, мы хотим, чтобы высота нашего дерева была равна `min(container height, total calculated tree height)`.
> 
> Одно из ограничений, которое трудно обойти, заключается в том, что мы не можем легко получить доступ к вычисленной высоте дерева без изменения кода реакции-арбориста. Однако мы можем добраться до него, опросив его в компоненте «Дерево» — к сожалению, нет другого хорошего способа выяснить, КОГДА это станет доступно в DOM. Это было бы намного проще с кодом, который позволял бы вам подписываться на изменения общей рассчитанной высоты дерева!
> 
> Я экстраполирую свою рабочую версию, чтобы показать принципы работы:
> 
> Сначала настройте DOM и несколько ссылок.
> 
> ```
> const screenRef = useRef(); // where you want to fill the available space of
> const containerRef = useRef(); // A div immediately above the tree (may be able to skip this actually (?) and do everything off the screenRef)
> const [treeHeight, setTreeHeight] = useState();
> <div css={css`height: 100vh; display: flex; flex-direction: column}` ref={screenRef}>
>   <div css={css`height: fit-content;`} ref={containerRef}>
>     <Tree height={treeHeight} data={...}>
>   </div>
> </div>
> ```
> 
> Тогда давайте представим код опроса.
> 
> ```
> // 500 = arbitrary starting height
> const [treeContentsHeight, setTreeContentsHeight] = useState(500);
> const [screenHeight, setScreenHeight] = useState(500);
> useEffect(() => {
>     // This observer 
>     const observer = new ResizeObserver(entries => {
>       const height = entries[0].contentRect.height;
>       setTreeContentsHeight(height);
>     });
>     // This tree has the correct total height calculated so we want to watch it using ResizeObserver
>     let innerTreeElement;
>     let pollId;
>     const poll = () => {
>       pollId = setTimeout(() => {
>         innerTreeElement = containerRef.current?.firstChild?.firstChild?.firstChild;
>         if (innerTreeElement) {
>           clearTimeout(pollId);
>           observer.observe(innerTreeElement);
>         } else {
>           pollId = poll();
>         }
>       }, 30);
>     };
>     poll();
>     return () => {
>       observer.disconnect();
>       clearTimeout(pollId);
>     };
>   }, []);
> ```
> 
> Наконец, всякий раз, когда TreeContentsHeight обновляется, мы можем установитьContainerHeight и убедиться, что он также обновляется при изменении размера.
> 
> ```
>   useEffect(() => {
>     const resizeContainer = (): void => {
>       const screenRefHeight = screenRefHeight.current.getBoundingClientRect().height;
>       setTreeHeight(Math.min(screenRefHeight, treeContentsHeight));
>     };
>     resizeContainer();
>     window.addEventListener('resize', resizeContainer);
>     return () => window.removeEventListener('resize', resizeContainer);
>   }, [treeContentsHeight]);
> ```

You are simply a genius! Thank you, your method helped me implement the dynamic height of the tree container.
In my case, height 100vh was drawing extra space after Tree , but I set height to treeContentsHeight and the height actually became dynamic

**Comment by sauron918** [comment]
I've followed the recommendations for using `useResizeObserver`, but I'm still facing issues with dynamic height.

<img src="https://github.com/user-attachments/assets/e8cdb9b2-8c0c-4161-9066-9a5b0b326ef1" width="200" />

Isn't it expected that the vertical scrollbar should appear when the page is resized? It's reproducible on https://react-arborist.netlify.app/gmail.

Notably, height is being adjusted only for height increases and only after 500px.

<img src="https://github.com/user-attachments/assets/40a8d3ce-69a1-4ee3-b746-ce466c05cb38" width="300" />

**Comment by jonahallibone** [comment]
Is it not possible to have the scrollbar hidden when the content is not expanded? The height of my tree is 2000px but the parent is only 1163px, meaning there's a pretty enormous gap at all times on the screen?

**Comment by jorgedanisc** [comment]
> It looks like it could be possible to use custom scrollbars if the `outerElementType` of the `react-window` implementation was able to be swapped out: [bvaughn/react-window#110 (comment)](https://github.com/bvaughn/react-window/issues/110#issuecomment-469061213).
> 
> In the example they are using `react-custom-scrollbars` so I don't see why another component such as `ScrollArea` from Radix shouldn't work.

As mentioned here, would it be possible to provide a simple Element prop from the `<Tree />` to modify the `outerElementType` from the `<FixedSizeList>` ?

https://github.com/brimdata/react-arborist/blob/c851c6f76ad5878f3ea7b68602e9b2f970df142e/modules/react-arborist/src/components/list-outer-element.tsx

This would be a huge feature for somebody that needs to have a custom scroll bar on critical layout shift control scenarios. Like mentioned above on the quote, a shadcn/radix `<ScrollArea>` or else could be used here if need be
